grails-app/services/org/pih/warehouse/product/ProductService.groovy

```diff
@@ -1411,4 +1411,15 @@ class ProductService {
         }
         return product
     }
+
+    Synonym editProductSynonym(String synonymId, String synonymTypeCodeName, String synonymValue, String localeName) {
+        Synonym synonym = Synonym.get(synonymId)
+        Locale locale = localeName ? new Locale(localeName) : null
+        SynonymTypeCode synonymTypeCode = synonymTypeCodeName ? SynonymTypeCode.valueOf(synonymTypeCodeName) : SynonymTypeCode.ALTERNATE_NAME
+        synonym.properties = [ locale: locale, synonymTypeCode: synonymTypeCode, name: synonymValue ]
+        if (!synonym.validate() || !synonym.save(flush: true, failOnError: true)) {
+            throw new ValidationException("Invalid synonym", synonym.errors)
```

Comment:
`failOnError:true` seems like overkill here since both validate() and save() should lead us to a validation exception. failOnError will probably throw it's own ValidationException so we just need to make sure that both lead to the same outcome. I'm not sure how to test that out though.